File: backend/internal/simulation/clock.go
package simulation

import (
	"sync"
	"time"
)

// VirtualClock provides deterministic time control for simulation testing
// Inspired by FoundationDB's simulation clock and TigerBeetle's VOPR
type VirtualClock struct {
	mu          sync.RWMutex
	current     time.Time
	timers      []*VirtualTimer
	nextTimerID int
}

// NewVirtualClock creates a clock starting at the given time
func NewVirtualClock(start time.Time) *VirtualClock {
	return &VirtualClock{
		current: start,
		timers:  make([]*VirtualTimer, 0, 100),
	}
}

// Now returns the current simulated time
func (c *VirtualClock) Now() time.Time {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.current
}

// Advance moves time forward by the given duration
// Returns slice of timers that fired during this advance
func (c *VirtualClock) Advance(d time.Duration) []*VirtualTimer {
	c.mu.Lock()
	defer c.mu.Unlock()

	c.current = c.current.Add(d)
	fired := make([]*VirtualTimer, 0)

	// Check which timers should fire
	remaining := make([]*VirtualTimer, 0, len(c.timers))
	for _, timer := range c.timers {
		if !timer.deadline.After(c.current) {
			// Timer fired
			fired = append(fired, timer)
			timer.Fire()
		} else {
			// Timer still pending
			remaining = append(remaining, timer)
		}
	}

	c.timers = remaining
	return fired
}

// After schedules a callback to execute after the given duration
// Returns a VirtualTimer that can be cancelled
func (c *VirtualClock) After(d time.Duration, callback func()) *VirtualTimer {
	c.mu.Lock()
	defer c.mu.Unlock()

	timer := &VirtualTimer{
		id:       c.nextTimerID,
		deadline: c.current.Add(d),
		callback: callback,
		clock:    c,
	}
	c.nextTimerID++

	c.timers = append(c.timers, timer)
	return timer
}

// AfterFunc schedules a callback (time.AfterFunc compatible)
func (c *VirtualClock) AfterFunc(d time.Duration, f func()) *VirtualTimer {
	return c.After(d, f)
}

// Sleep advances time (simulated sleep - instant in real time)
func (c *VirtualClock) Sleep(d time.Duration) {
	c.Advance(d)
}

// PendingTimers returns the number of scheduled timers
func (c *VirtualClock) PendingTimers() int {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return len(c.timers)
}

// Reset resets the clock to the given time and cancels all timers
func (c *VirtualClock) Reset(t time.Time) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.current = t
	c.timers = make([]*VirtualTimer, 0, 100)
}

// VirtualTimer represents a scheduled callback in simulated time
type VirtualTimer struct {
	id       int
	deadline time.Time
	callback func()
	clock    *VirtualClock
	fired    bool
	mu       sync.Mutex
}

// Fire executes the timer callback (called by clock)
func (t *VirtualTimer) Fire() {
	t.mu.Lock()
	defer t.mu.Unlock()

	if t.fired {
		return // Already fired
	}

	t.fired = true
	if t.callback != nil {
		go t.callback() // Execute in goroutine to simulate async
	}
}

// Cancel prevents the timer from firing
func (t *VirtualTimer) Cancel() bool {
	t.mu.Lock()
	defer t.mu.Unlock()

	if t.fired {
		return false
	}

	// Remove from clock's timer list
	t.clock.mu.Lock()
	defer t.clock.mu.Unlock()

	remaining := make([]*VirtualTimer, 0, len(t.clock.timers))
	for _, timer := range t.clock.timers {
		if timer.id != t.id {
			remaining = append(remaining, timer)
		}
	}
	t.clock.timers = remaining

	t.fired = true
	return true
}

// Stop is an alias for Cancel (time.Timer compatible)
func (t *VirtualTimer) Stop() bool {
	return t.Cancel()
}

File: packages/simulation/src/clock.ts
/**
 * VirtualClock provides deterministic time control for simulation testing
 *
 * Inspired by FoundationDB's simulation clock and TigerBeetle's VOPR.
 * All time progression is controlled by the test, enabling instant time advancement
 * and perfect reproducibility.
 *
 * @example
 * ```ts
 * const clock = new VirtualClock(new Date('2025-01-01'));
 *
 * // Schedule callbacks
 * clock.setTimeout(() => console.log('fired!'), 1000);
 *
 * // Advance time instantly
 * clock.advance(1000); // Fires callback immediately
 * ```
 */
export class VirtualClock {
  private current: Date;
  private timers: VirtualTimer[] = [];
  private nextTimerID = 0;
  private intervals: Map<number, { callback: () => void; intervalMs: number; nextFire: Date }> = new Map();

  constructor(startTime: Date = new Date('2025-01-01T00:00:00Z')) {
    this.current = new Date(startTime);
  }

  /**
   * Returns the current simulated time
   */
  now(): Date {
    return new Date(this.current);
  }

  /**
   * Advances time by the given milliseconds
   * Fires all timers and intervals that should trigger during this period
   *
   * @returns Array of timer IDs that fired
   */
  advance(ms: number): number[] {
    const targetTime = new Date(this.current.getTime() + ms);
    const firedIDs: number[] = [];

    // Fire timers
    const remainingTimers: VirtualTimer[] = [];
    for (const timer of this.timers) {
      if (timer.deadline <= targetTime) {
        firedIDs.push(timer.id);
        timer.callback();
      } else {
        remainingTimers.push(timer);
      }
    }
    this.timers = remainingTimers;

    // Fire intervals
    for (const [id, interval] of this.intervals) {
      while (interval.nextFire <= targetTime) {
        interval.callback();
        interval.nextFire = new Date(interval.nextFire.getTime() + interval.intervalMs);
        firedIDs.push(id);
      }
    }

    this.current = targetTime;
    return firedIDs;
  }

  /**
   * Schedules a callback to execute after the given delay
   * Compatible with standard setTimeout signature
   */
  setTimeout(callback: () => void, ms: number): VirtualTimer {
    const timer: VirtualTimer = {
      id: this.nextTimerID++,
      deadline: new Date(this.current.getTime() + ms),
      callback,
    };
    this.timers.push(timer);
    return timer;
  }

  /**
   * Schedules a callback to execute repeatedly at the given interval
   * Compatible with standard setInterval signature
   */
  setInterval(callback: () => void, ms: number): number {
    const id = this.nextTimerID++;
    this.intervals.set(id, {
      callback,
      intervalMs: ms,
      nextFire: new Date(this.current.getTime() + ms),
    });
    return id;
  }

  /**
   * Cancels a timeout created by setTimeout
   */
  clearTimeout(timer: VirtualTimer): void {
    this.timers = this.timers.filter(t => t.id !== timer.id);
  }

  /**
   * Cancels an interval created by setInterval
   */
  clearInterval(id: number): void {
    this.intervals.delete(id);
  }

  /**
   * Returns the number of pending timers
   */
  pendingTimers(): number {
    return this.timers.length + this.intervals.size;
  }

  /**
   * Clears all timers and intervals
   */
  reset(): void {
    this.timers = [];
    this.intervals.clear();
  }
}

export interface VirtualTimer {
  id: number;
  deadline: Date;
  callback: () => void;
}
